### PR TITLE
[bugfix] Fixes attribute table duplicated rows #15974

### PR DESCRIPTION
**Pull request by elpaso**
The problem seems to be that the filter model is not in sync with the master (attribute table) model after the rows has been removed (that happens after the commitChanges when the provider has got its final feature IDs and the rows with the temporary IDs must be removed).

The solution proposed here is probably nothing more than a patch.

BTW, a C++ test has been added for the regression and all the changes in the code are necessary to make the test pass without segfaulting.

### CODE REVIEW

File: src/gui/layertree/qgslayertreeview.cpp [hunk 1]
```diff
@@ -253,20 +253,22 @@ void QgsLayerTreeView::updateExpandedStateFromNode( QgsLayerTreeNode *node )
 
 QgsMapLayer *QgsLayerTreeView::layerForIndex( const QModelIndex &index ) const
 {
-  QgsLayerTreeNode *node = layerTreeModel()->index2node( index );
-  if ( node )
+  if ( index.isValid( ) )
   {
```
Comment (nyalldawson) [inline src/gui/layertree/qgslayertreeview.cpp]: Are these changes related or included by accident?

Comment (elpaso) [inline src/gui/layertree/qgslayertreeview.cpp]: These changes prevent a segfault that I run into when developing the test, I'm not sure that it will be happen in the real world use case, but better stay on the safe side.

The reason was that the model was not set, in that case the index was also invalid, In fact I was thinking to also check the model in addition to (or instead of) the check on the index validity:
```
if ( layerTreeModel() && index.isValid( ) )
{
...
}
```

File: src/gui/attributetable/qgsattributetablemodel.cpp [hunk 2]
```diff
@@ -225,8 +227,12 @@ void QgsAttributeTableModel::featureAdded( QgsFeatureId fid )
     int n = mRowIdMap.size();
     beginInsertRows( QModelIndex(), n, n );
 
-    mIdRowMap.insert( fid, n );
-    mRowIdMap.insert( n, fid );
+    // Do not add if already there!
+    if ( ! mIdRowMap.contains( fid ) )
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetablemodel.cpp]: I think the `begin-` and `endInsertRows` should also be within this `if` block.

File: src/gui/attributetable/qgsdualview.cpp [hunk 3]
```diff
@@ -302,6 +303,11 @@ void QgsDualView::initModels( QgsMapCanvas *mapCanvas, const QgsFeatureRequest &
 
   mFilterModel = new QgsAttributeTableFilterModel( mapCanvas, mMasterModel, mMasterModel );
 
+  // The following connections to invalidate() are necessary to keep the filter model in sync
+  // see regression #15974
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsdualview.cpp]: better insert a link here ;)

Comment (m-kuhn) [inline src/gui/attributetable/qgsdualview.cpp]: @nyalldawson will these affect your performance improvements?

**Review by m-kuhn — COMMENTED** [review]
I don't think there's any danger involved in this PR (apart from the performance, to be confirmed by @nyalldawson ).
Since I still don't know what has gone wrong, I can't say if it's the proper fix.
Therefore it's good to merge for me after the comments have been taken into consideration, if it helps to fix issues.
